### ISSUE TITLE
FE_NedelecSZ does not compile

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#6557]
Something broke recently, but I don't know whether it was the merge of the NedelecSZ patch (#2240). Either way, all three resters
https://tjhei.info/cib/dealii-tjhei-osx/e86ca10dccde2c01d36694b3405a4bfebb7ad5a4/log
http://www.math.clemson.edu/~heister/cib/dealii-tjhei-gcc-serial/e86ca10dccde2c01d36694b3405a4bfebb7ad5a4/log-gcc
http://www.math.clemson.edu/~heister/cib/dealii-tjhei-alpha/e86ca10dccde2c01d36694b3405a4bfebb7ad5a4/log-clang
report error like this:
```
FAILED: source/fe/CMakeFiles/obj_fe_debug.dir/fe_nedelec_sz.cc.o 
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++  -DDEBUG -DTBB_DO_ASSERT=1 -DTBB_USE_DEBUG -Isource/fe -Iinclude -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/tbb-2018_U2/include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/boost-1.62.0/include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/umfpack/UMFPACK/Include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/umfpack/AMD/Include -I/Users/cib-osx/auto-dealii-tjhei-osx/dealii/bundled/muparser_v2_2_4/include -pedantic -fPIC -Wall -Wextra -Wpointer-arith -Wwrite-strings -Wsynth -Wsign-compare -Wswitch -Woverloaded-virtual -Wno-deprecated-declarations -Qunused-arguments -Wno-unsupported-friend -Wno-undefined-var-template -openmp-simd -std=c++1z -ftemplate-depth=1024 -Werror -Wno-parentheses -Wno-unused-local-typedefs -O0 -ggdb -Wa,--compress-debug-sections -MD -MT source/fe/CMakeFiles/obj_fe_debug.dir/fe_nedelec_sz.cc.o -MF source/fe/CMakeFiles/obj_fe_debug.dir/fe_nedelec_sz.cc.o.d -o source/fe/CMakeFiles/obj_fe_debug.dir/fe_nedelec_sz.cc.o -c /Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/fe/fe_nedelec_sz.cc
/Users/cib-osx/auto-dealii-tjhei-osx/dealii/source/fe/fe_nedelec_sz.cc:882:10: error: no viable conversion from returned value of type 'unique_ptr<dealii::FE_NedelecSZ<2>::InternalData>' to function return type 'unique_ptr<typename FiniteElement<2, 2>::InternalDataBase>'
  return data;
         ^~~~
source/fe/fe_nedelec_sz.inst:17:17: note: in instantiation of member function 'dealii::FE_NedelecSZ<2>::get_data' requested here
 template class FE_NedelecSZ< 2 >;
                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2565:29: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::unique_ptr<InternalData>' to 'const std::__1::unique_ptr<dealii::FiniteElement<2, 2>::InternalDataBase, std::__1::default_delete<dealii::FiniteElement<2, 2>::InternalDataBase> > &' for 1st argument
class _LIBCPP_TYPE_VIS_ONLY unique_ptr
                            ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2594:49: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<InternalData>' to 'nullptr_t' for 1st argument
    _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR unique_ptr(nullptr_t) _NOEXCEPT
                                                ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2621:31: note: candidate constructor not viable: no known conversion from 'std::unique_ptr<InternalData>' to 'std::__1::unique_ptr<dealii::FiniteElement<2, 2>::InternalDataBase, std::__1::default_delete<dealii::FiniteElement<2, 2>::InternalDataBase> > &&' for 1st argument
    _LIBCPP_INLINE_VISIBILITY unique_ptr(unique_ptr&& __u) _NOEXCEPT
                              ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2625:9: note: candidate constructor [with _Up = dealii::FE_NedelecSZ<2>::InternalData, _Ep = std::__1::default_delete<dealii::FE_NedelecSZ<2>::InternalData>] not viable: no known conversion from 'std::unique_ptr<InternalData>' to 'unique_ptr<dealii::FE_NedelecSZ<2>::InternalData, std::__1::default_delete<dealii::FE_NedelecSZ<2>::InternalData> > &&' for 1st argument
        unique_ptr(unique_ptr<_Up, _Ep>&& __u,
        ^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:2640:35: note: candidate template ignored: could not match 'auto_ptr' against 'unique_ptr'
        _LIBCPP_INLINE_VISIBILITY unique_ptr(auto_ptr<_Up>&& __p,
                                  ^
```
and
```
/home/bob/source/source/fe/fe_nedelec_sz.cc: In instantiation of â€˜std::unique_ptr<typename dealii::FiniteElement<dim, dim>::InternalDataBase> dealii::FE_NedelecSZ<dim>::get_data(dealii::UpdateFlags, const dealii::Mapping<dim, dim>&, const dealii::Quadrature<dim>&, dealii::internal::FEValuesImplementation::FiniteElementRelatedData<dim, dim>&) const [with int dim = 2; typename dealii::FiniteElement<dim, dim>::InternalDataBase = dealii::FiniteElement<2, 2>::InternalDataBase]â€™:
source/fe/fe_nedelec_sz.inst:17:17:   required from here
/home/bob/source/source/fe/fe_nedelec_sz.cc:882:10: error: could not convert â€˜dataâ€™ from â€˜std::unique_ptr<dealii::FE_NedelecSZ<2>::InternalData, std::default_delete<dealii::FE_NedelecSZ<2>::InternalData> >â€™ to â€˜std::unique_ptr<dealii::FiniteElement<2, 2>::InternalDataBase, std::default_delete<dealii::FiniteElement<2, 2>::InternalDataBase> >â€™
   return data;
          ^
```
I suspect we missed this because the branch was so old that the testers were testing against the root of that  branch, but that that is no longer compatible with current master.

I'll take a look next.

**Comment by kronbichler** [comment]
I must admit that #2240 was a bit strange regarding the testers and would report that there was no permission, so I checked that locally on my machine and confirmed it worked there (to not let the authors wait even longer). So yes, the situation is strange and I actually don't see how they would file, as it complains that it cannot convert `unique_ptr<FE_NedelecSZ::InternalData>` into `unique_ptr<FiniteElement::InternalDataBase>` - and the former is derived from the latter, a pattern we use elsewhere if I remember correctly.

**Comment by bangerth** [comment]
I figured it out in the meantime. It's a compiler version problem, and we've had that before in a couple of patches.

**Comment by masterleinad** [comment]
I just added the fix.

**Comment by bangerth** [comment]
Ah, yes, we came to the same conclusion.

Someone will have to decide whether to take it from #6556 or #6558. Or merge both in order of #6558 then #6556 -- they are the same, so we should end up with two patches for the same problem :-)

**Comment by kronbichler** [comment]
Fixed by #6556.
